java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java

```diff
@@ -36,18 +37,41 @@
 
 public class ExecutionState {
 
-  final ExecutionState parent;
-  private SetMultimap<Symbol, SymbolicValue> reachableValues = HashMultimap.create();
+  public final ExecutionState parent;
+
+  /*
+   * definitelyReachableValues and potentiallyReachableValues contains the set of possible values from a given symbol.
+   * - definitelyReachableValues contains the values that are definitely retrieved by a read operation.
+   * - potentiallyReachableValues contains all the values that can be retrieved by a read operation.
+   * this is illustrated in the examples below:
+   * example 1                      example 2                        example 3
+   * a = V1;                        a = V1;                          a = V1;
+   * if(condition) {                if(condition) {                  a = V2;
+   *   a = V2;                        a = V2;
+   * } else {                       }
+   *   a = V3;
+   * }
+   * a.hashCode();                  a.hashCode();                    a.hashCode();
+   * a.unlock();                    a.unlock();                      a.unlock();
+   * definitely reachable: V2, V3   definitely reachable: V2         definitely reachable: V2
+   * potentially reachable: V2, V3  potentially reachable: V1, V2    potentially reachable: V2
+   */
+
+  // all values that are definitely modifiable (i.e. through an update).
+  private SetMultimap<Symbol, SymbolicValue> definitelyReachableValues = HashMultimap.create();
+  // all values that are reachable (i.e. through a read).
+  private final SetMultimap<Symbol, SymbolicValue> potentiallyReachableValues = HashMultimap.create();
   private SetMultimap<Symbol, SymbolicValue> unreachableValues = HashMultimap.create();
   /**
    * List of symbol that were declared within this execution state.
    */
+  // FIXME(merciesa): this rather relates to scope and should eventually be moved to the visitor.
```

Comment:
![New](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/new.png) ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)